### PR TITLE
MTDSA-6763 v2 empty body error

### PR DESCRIPTION
**Pull request by yangwong9664**

### CODE REVIEW

File: app/v2/controllers/requestParsers/validators/validations/JsonFormatValidation.scala [hunk 1]
```diff
@@ -16,18 +16,48 @@
...
+      case (path: JsPath, Seq(JsonValidationError(Seq(error: String)))) if error.contains("error.expected") => WrongFieldType(path)
+      case (path: JsPath, _) => OtherFailure(path)
     }
 
+    val logString = failures.groupBy(_.getClass)
+      .values.map(failure => s"${failure.head.failureReason}: " + s"${failure.map(_.fromJsPath)}")
+      .toString().dropRight(1).drop(5)
+
+    Logger.warn(s"[JsonFormatValidation][validate] - Request body failed validation with errors - $logString")
+    List(RuleIncorrectOrEmptyBodyError)
```
Comment (james-work-account) [inline app/v2/controllers/requestParsers/validators/validations/JsonFormatValidation.scala:47]: You're just returning RuleIncorrectOrEmptyBodyError with no paths - what's the point of this change? Just to add logs? If the point is to return paths too, please add tests for the new functionality.

Comment (yangwong9664) [reply app/v2/controllers/requestParsers/validators/validations/JsonFormatValidation.scala]: Missed adding the path change here, now added with tests.

File: app/v2/controllers/requestParsers/validators/TriggerBSASValidator.scala [hunk 2]
```diff
@@ -38,28 +36,26 @@ class TriggerBSASValidator @Inject()(val currentDateProvider: CurrentDateProvide
...
     List(
-      JsonValidation.validate(data.body.json \ "selfEmploymentId")(SelfEmploymentIdValidation.validate)
+      SelfEmploymentIdValidation.validateOption(data.body.json.\("selfEmploymentId").asOpt[String])
     )
   }
 
   private def typeOfBusinessValidator: TriggerBsasRawData => List[List[MtdError]] = { data =>
     List(
-      JsonValidation.validate(data.body.json \ "typeOfBusiness")(TypeOfBusinessValidation.validate)
+      TypeOfBusinessValidation.validateOption(data.body.json.\("typeOfBusiness").asOpt[String]),
```
Comment (james-work-account) [inline app/v2/controllers/requestParsers/validators/TriggerBSASValidator.scala:58]: I really don't like how the validation is done in here... normally, we'd have a `parameterFormatValidation`, `parameterRuleValidation`, `incorrectOrEmptyBodyValidation`, `bodyFormatValidation` and `bodyRuleValidation` (or something like that). Having all of these as separate validations (`dateFieldValidator`, `selfEmploymentIdValidator` etc) is kinda gross, it's making it much harder to tell what's going on in here. I'll go through and refactor some other time, just making a note that this is naaaasty.

Comment (james-work-account) [reply app/v2/controllers/requestParsers/validators/TriggerBSASValidator.scala:58]: Also, please replace `.\(...)`

File: app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala [hunk 3]
```diff
@@ -16,18 +16,48 @@
...
+import play.api.Logger
+import play.api.libs.json._
+import v2.models.errors.{MtdError, RuleIncorrectOrEmptyBodyError}
 
 /**
-  * Utilities to assist using validations where the value to validate comes from a JSON element
-  */
+ * Utilities to assist using validations where the value to validate comes from a JSON element
+ */
 object JsonValidation {
```
Comment (james-work-account) [inline app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala:26]: This file has 0% coverage.

File: app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala [hunk 4]
```diff
@@ -16,18 +16,48 @@
...
+    val logString = failures.groupBy(_.getClass)
+      .values.map(failure => s"${failure.head.failureReason}: " + s"${failure.map(_.fromJsPath)}")
+      .toString().dropRight(1).drop(5)
+
+    Logger.warn(s"[JsonFormatValidation][validate] - Request body failed validation with errors - $logString")
+    List(RuleIncorrectOrEmptyBodyError.copy(paths = Some(failures.map(_.fromJsPath))))
   }
+
+  private class JsonFormatValidationFailure(path: JsPath, failure: String) {
+    val failureReason: String = this.failure
```
Comment (james-work-account) [inline app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala:52]: `this.` is unnecessary

File: app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala [hunk 5]
```diff
@@ -16,18 +16,48 @@
...
+
+    def fromJsPath: String = this.path
```
Comment (james-work-account) [inline app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala:54]: `this.` is unnecessary

File: app/v2/controllers/requestParsers/validators/validations/JsonFormatValidation.scala [hunk 6]
```diff
@@ -16,18 +16,47 @@
 
 package v2.controllers.requestParsers.validators.validations
 
+import play.api.Logger
 import play.api.libs.json._
-import v2.models.errors.MtdError
+import v2.models.errors.{MtdError, RuleIncorrectOrEmptyBodyError}
 
 object JsonFormatValidation {
```
Comment (james-work-account) [inline app/v2/controllers/requestParsers/validators/validations/JsonFormatValidation.scala:23]: This file has 0% coverage.

File: app/v1/controllers/requestParsers/validators/Validator.scala [hunk 7]
```diff
@@ -37,3 +37,22 @@ trait Validator[A <: RawData] {
     }
   }
 }
+
+object Validator {
+
+  def flattenErrors(errors: List[List[MtdError]]): List[MtdError] = {
```
[the remark this answers is not in the document]
Comment (yangwong9664) [reply app/v1/controllers/requestParsers/validators/Validator.scala:43]: I was just copying what was in the initial proof of concept PR, but I can change this if you like

File: app/v2/controllers/requestParsers/validators/TriggerBSASValidator.scala [hunk 8]
```diff
@@ -38,28 +36,26 @@ class TriggerBSASValidator @Inject()(val currentDateProvider: CurrentDateProvide
...
-      JsonValidation.validate(endDate)(DateValidation.validate(EndDateFormatError))
+      DateValidation.validateOpt(StartDateFormatError)((data.body.json \ "accountingPeriod" \ "startDate").asOpt[String]),
+      DateValidation.validateOpt(EndDateFormatError)((data.body.json \ "accountingPeriod" \ "endDate").asOpt[String])
     )
   }
 
   private def selfEmploymentIdValidator: TriggerBsasRawData => List[List[MtdError]] = { data =>
     List(
-      JsonValidation.validate(data.body.json \ "selfEmploymentId")(SelfEmploymentIdValidation.validate)
+      SelfEmploymentIdValidation.validateOption(data.body.json.\("selfEmploymentId").asOpt[String])
```
[the remark this answers is not in the document]
Comment (yangwong9664) [reply app/v2/controllers/requestParsers/validators/TriggerBSASValidator.scala:52]: they both do the same thing, but can change if you like

File: app/v2/controllers/requestParsers/validators/TriggerBSASValidator.scala (hunk 2 above)
Comment (yangwong9664) [reply app/v2/controllers/requestParsers/validators/TriggerBSASValidator.scala:58]: Okay a refactor would be beneficial. Also`.\` does the same as ` \ ` but will change if that's an issue

File: app/v1/controllers/requestParsers/validators/Validator.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (BetaDraconis) [reply app/v1/controllers/requestParsers/validators/Validator.scala:43]: This is a relic from an older version of the POC where missing mandatory field and incorrect field type errors were separate and distinct, and had to be flattened in the validation.

File: app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala (hunk 3 above)
Comment (yangwong9664) [reply app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala:26]: This file is unused anywhere in the application, thanks for finding

File: app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala (hunk 4 above)
Comment (yangwong9664) [reply app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala:52]: This was copied from the original POC commit, if this is an issue then it needs to be raised on all sub-tasks on MTDSA-6531. Will remove from this branch.

File: app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala (hunk 5 above)
Comment (yangwong9664) [reply app/v2/controllers/requestParsers/validators/validations/JsonValidation.scala:54]: This was copied from the original POC commit, if this is an issue then it needs to be raised on all sub-tasks on MTDSA-6531. Will remove from this branch.

File: app/v2/controllers/requestParsers/validators/validations/JsonFormatValidation.scala (hunk 6 above)
Comment (yangwong9664) [reply app/v2/controllers/requestParsers/validators/validations/JsonFormatValidation.scala:23]: This file is unused anywhere in the application, thanks for finding

**Pull request merged**